### PR TITLE
Call onTapCancel when down pointer gets cancelled

### PR DESCRIPTION
**Pull request by goderbauer**
## Description

Properly execute onTapCancel when a pointer gets cancelled after a down event has been received. This caused ink splashes to never go away if a pointer down event was not followed by a pointer up event, see https://github.com/flutter/flutter/pull/28533#issue-256469306 for more context.

Also includes a fix where onTapCancel got called even though onTapDown was never called. That was violating documented behavior: https://master-docs.flutter.io/flutter/gestures/TapGestureRecognizer/onTapCancel.html

This PR was salvaged from https://github.com/flutter/flutter/pull/28533.

## Related Issues

None

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes tests for *all* changed/updated/fixed behaviors (See [Test Coverage]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/gestures/tap_test.dart [hunk 1]
```diff
@@ -398,10 +397,75 @@ void main() {
...
+    tester.closeArena(5);
+    tester.route(down);
+    expect(recognized, <String>['down']);
+    tester.route(cancel);
+    expect(recognized, <String>['down', 'cancel']);
+
+    tap.dispose();
+  });
+
+  testGesture('loosing tap gesture recognizer does not send onTapCancel', (GestureTester tester) {
```
Comment (xster) [inline packages/flutter/test/gestures/tap_test.dart]: losing

**Comment by xster** [comment]
LGTM

**Comment by goderbauer** [comment]
I had to update some other tests that were expecting a onTapCancel without ever getting a onTapDown, which is violating the documented contract: http://docs.flutter.io/flutter/gestures/TapGestureRecognizer/onTapCancel.html

I believe the new behavior is more correct, but PTAL.

File: packages/flutter/test/widgets/gesture_detector_test.dart [hunk 2]
```diff
@@ -299,21 +299,21 @@ void main() {
       await gesture.up();
```
Comment (xster) [inline packages/flutter/test/widgets/gesture_detector_test.dart]: fix comments in 290 and 295

**Comment by xster** [comment]
LGTM

File: packages/flutter/test/gestures/tap_test.dart [hunk 3]
```diff
@@ -398,10 +397,75 @@ void main() {
...
       'tapA onTapUp',
       'tapA onTap',
-      'tapB onTapCancel',
       'swept 2',
       'disposed A',
       'disposed B',
     ]);
   });
+
+  testGesture('PointerCancelEvent cancels tab', (GestureTester tester) {
```
Comment (HansMuller) [inline packages/flutter/test/gestures/tap_test.dart]: cancels "tap" or tap gesture

File: packages/flutter/test/gestures/tap_test.dart [hunk 4]
```diff
@@ -398,10 +397,75 @@ void main() {
...
+    tap.onTap = () {
+      recognized.add('tap');
+    };
+    tap.onTapCancel = () {
+      recognized.add('cancel');
+    };
+
+    tap.addPointer(down);
+    tester.closeArena(5);
+    tester.route(down);
```
Comment (HansMuller) [inline packages/flutter/test/gestures/tap_test.dart]: Why is this needed?

File: packages/flutter/test/widgets/text_selection_test.dart [hunk 5]
```diff
@@ -119,23 +119,23 @@ void main() {
     expect(singleLongTapStartCount, 0);
   });
 
-  testWidgets('a very quick swipe is just a canceled tap', (WidgetTester tester) async {
+  testWidgets('a very quick swipe is ignored', (WidgetTester tester) async {
```
Comment (HansMuller) [inline packages/flutter/test/widgets/text_selection_test.dart]: Are we testing a "quick swipe" or a swipe that doesn't begin with a pointer down event?

**Review by HansMuller — COMMENTED** [review]
This looks OK although not sure about not cancelling a gesture if we don't see the down event. What about about drags that start outside of the gesture recognizer? Do we always ignore them?

File: packages/flutter/test/gestures/tap_test.dart (hunk 4 above)
Comment (goderbauer) [reply packages/flutter/test/gestures/tap_test.dart]: Removed.

File: packages/flutter/test/widgets/text_selection_test.dart (hunk 5 above)
Comment (goderbauer) [reply packages/flutter/test/widgets/text_selection_test.dart]: It's a quick swipe with a pointer down event, but the pointer down event doesn't trigger a tapDown.

**Comment by goderbauer** [comment]
@HansMuller I confirmed that drags that start outside a TapGestureDetector do not fire any events.

**Review by HansMuller — APPROVED** [review]
LGTM

File: packages/flutter/test/gestures/tap_test.dart [hunk 6]
```diff
@@ -398,10 +397,74 @@ void main() {
...
+    };
+    tap.onTap = () {
+      recognized.add('tap');
+    };
+    tap.onTapCancel = () {
+      recognized.add('cancel');
+    };
+
+    tap.addPointer(down);
+    tester.closeArena(5);
```
Comment (HansMuller) [inline packages/flutter/test/gestures/tap_test.dart]: I'm confused about exactly why we're closing the arena and he canceling the route 5 seconds later. It would help if you could explain this part of the test in terms of user gestures and widget tree changes.

File: packages/flutter/test/gestures/tap_test.dart [hunk 7]
```diff
@@ -398,10 +397,74 @@ void main() {
...
+      recognized.add('up');
+    };
+    tap.onTap = () {
+      recognized.add('tap');
+    };
+    tap.onTapCancel = () {
+      recognized.add('cancel');
+    };
+
+    tap.addPointer(down3);
```
Comment (HansMuller) [inline packages/flutter/test/gestures/tap_test.dart]: I assume that pointer 3 in this test or pointer 5 in the previous isn't significant; could be any one pointer?

Otherwise: same comment as the previous test. This doesn't look wrong but it's hard to understand if it's right.
